Added support for the ArcticForCausalLM.

fairydreaming opened the pull request:
Fixes #6877

Contains the following changes:
- increases maximum number of experts from 60 to 128
- adds new tensor type FFN_NORM_EXP (for a normalization block before MoE that runs in parallel to the attention + FFN, see #6877 for details)
- introduces architecture-specific block mappings in gguf-py (details in #6877)
- adds new model type MODEL_10B_128x3_66B
- adds new ARCTIC architecture and a general support for models based on this architecture

Model files for testing: https://huggingface.co/sszymczyk/snowflake-arctic-instruct-GGUF

sszymczy committed:
Added support for the snowflake-arctic model.

sorasoras commented:
It's possible to only offload dense part of the model onto GPU

sszymczy committed:
Whitespace formatting fixes.

cebtenzzre commented on convert-hf-to-gguf.py:
```diff
@@ -1517,6 +1517,119 @@ def write_tensors(self):
+    model_arch = gguf.MODEL_ARCH.ARCTIC
+
+    def set_vocab(self):
+        self._set_vocab_llama_hf()
```
re: https://github.com/ggerganov/llama.cpp/issues/6877#issuecomment-2089168239, this should be:

```suggestion
     try: 
         self. _set_vocab_sentencepiece() 
     except FileNotFoundError: 
         self._set_vocab_llama_hf() 
```

The assertion exists because LlamaHfVocab was primarily written to convert HF "fast" tokenizers with a tokenizer.json. Since before it existed, "slow" sentencepiece tokenizers with a tokenizer.model have (almost?) always been converted using SentencePieceProcessor, which doesn't depend on HF transformers and directly preserves the token types and scores.

If you want to start converting slow tokenizers using HfVocab as well, I won't stop you, but in order to be consistent you'd have to remove all references to SentencePieceProcessor in the convert scripts, and make HF transformers a hard requirement for converting models with a Llama vocab. Otherwise, we'd be making an exception for this model for no clear reason.

fairydreaming replied:
My reason is that the official tokenizer.model file for snowflake-arctic-instruct contains wrong BOS and EOS tokens as confirmed in: https://huggingface.co/Snowflake/snowflake-arctic-instruct/discussions/12
That's why I used llama_hf vocab that reads tokens from json files instead. If there is a better solution for this I'm fully open to any suggestions.

fairydreaming replied:
@cebtenzzre What if I implement ArcticModel::set_vocab() myself like XverseForCausalLM did, is that acceptable?

fairydreaming replied:
@cebtenzzre I now load vocabulary with SentencePieceProcessor as you suggested and apply necessary token modifications based on added_tokens_decoder field from tokenizer_config.json.

sszymczy committed:
Read vocabulary for ArcticForCausalLM from sentencepiece model instead of HF tokenizer.
Add/redefine tokens accordingly to added_tokens_decoder from tokenizer_config.json

compilade commented on convert-hf-to-gguf.py:
```diff
@@ -1517,6 +1517,119 @@ def write_tensors(self):
+        self.gguf_writer.add_vocab_size(hparams["vocab_size"])
+        self.gguf_writer.add_rope_dimension_count(hparams["hidden_size"] // hparams["num_attention_heads"])
+
+    # Same as super class, but permuting q_proj, k_proj
+    def write_tensors(self):
```
This can be simplified with the changes from #7075 by overriding `modify_tensors` instead of `write_tensors`.

fairydreaming replied:
@compilade I applied changes from #7075 and #6511 to the ArcticModel in convert-hf.

sszymczy committed:
Moved ArcticModel to the end of the file.

sszymczy committed:
Applied changes from upstream PR: save memory with lazy evaluation #7075 (shameless copy from LlamaModel).

sszymczy committed:
Replaced prints with logger calls.

fairydreaming committed:
Merge branch 'ggerganov:master' into snowflake-arctic-clean

sszymczy committed:
Removed unnecessary method - LlamaModel.permute is used instead.

compilade commented on convert-hf-to-gguf.py:
```diff
@@ -2277,6 +2277,156 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
+            sys.exit(1)
+
+        # Read the whole vocabulary from the tokenizer.model file
+        tokenizer = SentencePieceProcessor(str(tokenizer_path))
+
+        vocab_size = self.hparams.get('vocab_size', tokenizer.vocab_size())
+
+        tokens: list[bytes] = [f"[PAD{i}]".encode("utf-8") for i in range(vocab_size)]
+        scores: list[float] = [-10000.0] * vocab_size
+        toktypes: list[int] = [SentencePieceTokenTypes.UNKNOWN] * vocab_size
+
+        for token_id in range(tokenizer.vocab_size()):
+
+            piece = tokenizer.id_to_piece(token_id)
+            text = piece.encode("utf-8")
+            score = tokenizer.get_score(token_id)
+
+            toktype = SentencePieceTokenTypes.NORMAL
+            if tokenizer.is_unknown(token_id):
+                toktype = SentencePieceTokenTypes.UNKNOWN
+            elif tokenizer.is_control(token_id):
+                toktype = SentencePieceTokenTypes.CONTROL
+            elif tokenizer.is_unused(token_id):
+                toktype = SentencePieceTokenTypes.UNUSED
+            elif tokenizer.is_byte(token_id):
+                toktype = SentencePieceTokenTypes.BYTE
```
Some changes from #7075 are missing:

```diff
diff --git a/convert-hf-to-gguf.py b/convert-hf-to-gguf.py
index 170dea06..5569ed03 100755
--- a/convert-hf-to-gguf.py
+++ b/convert-hf-to-gguf.py
@@ -2294,7 +2294,8 @@ class ArcticModel(Model):
             sys.exit(1)
 
         # Read the whole vocabulary from the tokenizer.model file
-        tokenizer = SentencePieceProcessor(str(tokenizer_path))
+        tokenizer = SentencePieceProcessor()
+        tokenizer.LoadFromFile(str(tokenizer_path))
 
         vocab_size = self.hparams.get('vocab_size', tokenizer.vocab_size())
 
@@ -2304,18 +2305,18 @@ class ArcticModel(Model):
 
         for token_id in range(tokenizer.vocab_size()):
 
-            piece = tokenizer.id_to_piece(token_id)
+            piece = tokenizer.IdToPiece(token_id)
             text = piece.encode("utf-8")
-            score = tokenizer.get_score(token_id)
+            score = tokenizer.GetScore(token_id)
 
             toktype = SentencePieceTokenTypes.NORMAL
-            if tokenizer.is_unknown(token_id):
+            if tokenizer.IsUnknown(token_id):
                 toktype = SentencePieceTokenTypes.UNKNOWN
-            elif tokenizer.is_control(token_id):
+            elif tokenizer.IsControl(token_id):
                 toktype = SentencePieceTokenTypes.CONTROL
-            elif tokenizer.is_unused(token_id):
+            elif tokenizer.IsUnused(token_id):
                 toktype = SentencePieceTokenTypes.UNUSED
-            elif tokenizer.is_byte(token_id):
+            elif tokenizer.IsByte(token_id):
                 toktype = SentencePieceTokenTypes.BYTE
 
             tokens[token_id] = text
```

This does the same thing but makes `pyright` not scream with errors.
This is because SentencePiece [defines the snake_case aliases with some dynamic code](https://github.com/google/sentencepiece/blob/5253649c81ee5403edae2be517dc2915c71d774e/python/src/sentencepiece/__init__.py#L1154-L1164), but some static type checkers like `pyright` don't understand it.

fairydreaming replied:
Thanks, should be fixed now.

sszymczy committed:
convert-hf : Corrected sentencepiece API calls.

sszymczy committed:
Minor fixes after merging.

sszymczy committed:
Merge branch 'ggerganov:master' into snowflake-arctic

fairydreaming commented on gguf-py/gguf/tensor_mapping.py:
```diff
@@ -393,12 +451,16 @@ def __init__(self, arch: MODEL_ARCH, n_blocks: int):
             self.mapping[tensor_name] = (tensor, tensor_name)
             for key in keys:
                 self.mapping[key] = (tensor, tensor_name)
+        if arch in self.arch_block_mappings_cfg:
+            block_mappings = self.arch_block_mappings_cfg[arch]
```
So the idea is to keep only "conflicting" block mappings in architecture-specific mappings and "non-conflicting" mappings in general mappings? I think using dict.update() is a better idea then. Mappings for ARCTIC arch would be shortened to:

```
    # architecture-specific block mappings
    arch_block_mappings_cfg: dict[MODEL_ARCH, dict[MODEL_TENSOR, tuple[str, ...]]] = {
        MODEL_ARCH.ARCTIC: {
            MODEL_TENSOR.FFN_NORM: (
                "model.layers.{bid}.residual_layernorm",
            ),
            MODEL_TENSOR.FFN_NORM_EXP: (
                "model.layers.{bid}.post_attention_layernorm",
            ),
        },
    }
```
while in the TensorNameMap init we would only have to add:
```
        if arch in self.arch_block_mappings_cfg:
            self.block_mappings_cfg.update(self.arch_block_mappings_cfg[arch])
```
What do you think?

compilade replied:
> So the idea is to keep only "conflicting" block mappings in architecture-specific mappings and "non-conflicting" mappings in general mappings?

Yes, exactly.

> What do you think?

I think using `dict.update()` would be good. My proposed approach would have made a copy of the dict, but you're right, updating in-place would work too and would be better, since the original `block_mappings_cfg` isn't used later on (I think?).

I agree with using `dict.update()` for this.

fairydreaming replied:
OK, done

sszymczy committed:
gguf-py : Moved non-conflicting block mappings from architecture-specific ARCTIC mappigs to general mappings.

compilade commented on gguf-py/gguf/constants.py:
```diff
@@ -181,6 +182,7 @@ class MODEL_TENSOR(IntEnum):
     SSM_A              = auto()
     SSM_D              = auto()
     SSM_OUT            = auto()
+    FFN_NORM_EXP       = auto()
```
Since the actual numbers associated to the enum values of `MODEL_TENSOR` don't really matter (their names (from `TENSOR_NAMES`) are used instead in GGUF), maybe `FFN_NORM_EXP` could be placed right before `FFN_GATE_EXP`, a bit like `FFN_NORM` is right before `FFN_GATE`, for consistency.

If this is changed, it should also be placed similarly in `TENSOR_NAMES` and `MODEL_TENSORS[MODEL.ARCTIC]` in `gguf-py/gguf/constants.py` as well as in the `llm_tensor` enum, the `LLM_TENSOR_NAMES` mapping, and the `llama_layer` struct (and maybe the `LLM_ARCH_ARCTIC` case in `llm_load_tensors`?) in `llama.cpp`.

fairydreaming replied:
I changed the order as requested, but in llama_layer struct the order is different, so I didn't touch it. In llm_load_tensors I think it was already in the requested order.

compilade reviewed (approved):
I did not test this (the model is quite big), but the code looks good to me. Nice work @fairydreaming!

sszymczy committed:
Reordered tensors for visual consistency.

sszymczy committed:
llama : Removed usage of bias tensors in LLM_ARCH_ARCTIC, as they are not present in released models.

sszymczy committed:
Reordered tensors for visual consistency.

fairydreaming commented:
I noticed that the arctic model doesn't use bias tensors, so I removed usage of bias tensors in the LLM_ARCH_ARCTIC-related code (they were all nulls anyway).

sszymczy committed:
llama : Replaced obsolete ggml_rope_custom() calls with ggml_rope_ext().

sszymczy committed:
Merge remote-tracking branch 'upstream/master' into snowflake-arctic-clean

ggerganov commented:
I haven't tested as well, but it seems good so feel free to merge

fairydreaming commented:
> I haven't tested as well, but it seems good so feel free to merge

@ggerganov I noticed that Snowflake changed the Arctic model 2 weeks ago. The commit says: "Fixes for GQA support" and num_key_value_heads in config.json changed value from 56 to 8, so I have redownload the model and check if it still works.

sszymczy committed:
Merge remote-tracking branch 'upstream/master' into snowflake-arctic-clean

sszymczy committed:
llama : fix whitespace formatting
